### PR TITLE
073 override parameters

### LINKED ISSUE #598: override certain information given in config file with cmd line parameters
Additional command line parameters that override parameters given in the config file are desireable; at least for the global learn mode and file input parameter. This would enable the user to switch learn mode on/off without editing the config file; or to quickly switch between training and test file via the cmd line.

### PR DESCRIPTION
**Pull request by dITySoftware**
# Make sure these boxes are signed before submitting your Pull Request -- thank you.

# Must haves
- [x] I have read and followed the contributing guide lines at https://github.com/ait-aecid/logdata-anomaly-miner/wiki/Git-development-workflow
- [x] Issues exist for this PR
- [x] I added related issues using the "Fixes #<issue-id>"-notations
- [x] This Pull-Requests merges into the "development"-branch

Fixes #598 

# Submission specific

- [ ] This PR introduces breaking changes
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed

# Describe changes:
- added code to allow overriding config_properties from the command line.

### CODE REVIEW

File: source/root/usr/lib/logdata-anomaly-miner/aminer.py [hunk 1]
```diff
@@ -214,6 +244,11 @@ def main():
     parser.add_argument('-r', '--remove', action='append', type=str, help='removes a specific persistence directory')
     parser.add_argument('-R', '--restore', type=str, help='restore a persistence backup')
     parser.add_argument('-f', '--from-begin', action='store_true', help='removes RepositioningData before starting the aminer')
+    parser.add_argument("--config-properties", metavar="KEY=VALUE", nargs='+',
+                        help="Set a number of config_properties by using key-value pairs (do not put spaces before or after the = sign). "
+                             "If a value contains spaces, you should define it with double quotes: 'foo=\"this is a sentence\". Note that "
+                             "values are always treated as strings. If values are already defined in the config_properties, the input "
+                             "types are converted to the ones already existing.")
```
Comment (landauermax) [inline source/root/usr/lib/logdata-anomaly-miner/aminer.py:251]: The last sentence could be misleading, especially since it is not clear what "converted" means in this context. I would prefer something like "Properties that are defined in the configuration file are overwritten when specified as command-line parameters."

File: source/root/usr/lib/logdata-anomaly-miner/aminer.py [hunk 2]
```diff
@@ -214,6 +244,11 @@ def main():
     parser.add_argument('-r', '--remove', action='append', type=str, help='removes a specific persistence directory')
     parser.add_argument('-R', '--restore', type=str, help='restore a persistence backup')
     parser.add_argument('-f', '--from-begin', action='store_true', help='removes RepositioningData before starting the aminer')
+    parser.add_argument("--config-properties", metavar="KEY=VALUE", nargs='+',
+                        help="Set a number of config_properties by using key-value pairs (do not put spaces before or after the = sign). "
+                             "If a value contains spaces, you should define it with double quotes: 'foo=\"this is a sentence\". Note that "
+                             "values are always treated as strings. If values are already defined in the config_properties, the input "
```
Comment (landauermax) [inline source/root/usr/lib/logdata-anomaly-miner/aminer.py:250]: If values are always treated as strings, is it then not possible to set numeric values such as the debug level? If they are converted (and I think the aminer should convert them into the expected data format and print an error if conversion is not possible), then this sentence about treating values as strings can be omitted.

File: source/root/usr/lib/logdata-anomaly-miner/aminer.py (hunk 1 above)
Comment (dITySoftware) [reply source/root/usr/lib/logdata-anomaly-miner/aminer.py:251]: I have written that every value is treated as string. the last sentence should describe that the type is changed to the already existing one in the config or left as string.

For example:
a=5
"a" does not exist in the config so it is evaluated as a = "5".
In the other case that "a" is defined as a = 600, then the new "a" would be a = 5 (integer).

How could i write that more understandable?

File: source/root/usr/lib/logdata-anomaly-miner/aminer.py (hunk 2 above)
Comment (dITySoftware) [reply source/root/usr/lib/logdata-anomaly-miner/aminer.py:250]: Well how do you know what type belongs to a specific config if it is not defined in the original one?

The stat level and debug level is another parameter that already exists in the aminer.
A realistic example would be to change the LogResourceList. I have forgotten this case, but added it now. So the LogResourceList is a must have parameter, so it is defined every time. From that we know that it's type is list and can convert the input to a list.

Another example would be the Core.PersistencePeriod. This one does not have to be defined and therefore we can not check the type of it. So if it is not defined, it would be treated as string; in the other case it is treated as integer.

Basically you can only CHANGE parameters and can not ADD new parameters if they are not of the type string.
I think it is ok if documented.

Comment (landauermax) [reply source/root/usr/lib/logdata-anomaly-miner/aminer.py:251]: Okay, I was not aware that it is not possible to add parameters that are not string. Why is that the case? It should be possible to derive the expected type from the default value of that parameter, e.g., I think the default value of the PeristencePeriod is set to 600.

Comment (dITySoftware) [reply source/root/usr/lib/logdata-anomaly-miner/aminer.py:251]: I guess i can add that where applicable. However this does not solve the problem as not all configurations have default values. Also is it worth it to add so many individual checks just to extend this?
We can also just state that the type is string by default and existing configs are cast to the right type.

I think we can never guarantee that every parameter can be converted to the right type in every configuration. Maybe we should make it really clear (i. e. with WARNING or something like that in the help output).
Also we could just stop the aminer if the parameter was not already defined in the config. (or at least print a warning that the parameter xxx was added to the config as string.

Please choose the best solution and I will implement it.

Comment (landauermax) [reply source/root/usr/lib/logdata-anomaly-miner/aminer.py:251]: Yes, I think printing a warning is the least we should do at this point.

Comment (dITySoftware) [reply source/root/usr/lib/logdata-anomaly-miner/aminer.py:251]: I have added exception handling when the variable is converted to other types and a warning if the variable is not set in the config and that it will be treated as string.

Comment (landauermax) [reply source/root/usr/lib/logdata-anomaly-miner/aminer.py:251]: Okay, I think thats an appropriate solution for now.

File: source/root/usr/lib/logdata-anomaly-miner/aminer.py (hunk 1 above)
Comment (landauermax) [reply source/root/usr/lib/logdata-anomaly-miner/aminer.py:252]: Resolved in the other comment.

**Review by whotwagner — APPROVED** [review]
